test: refactor out same-txid-diff-wtxid tx to reuse in other tests

stratospher opened the pull request:
It's useful to easily create transactions with same txid, different wtxid and valid witness for testing scenarios in other places in the codebase (ex: private broadcast connections, see https://github.com/bitcoin/bitcoin/pull/29415#discussion_r2055915118)

So refactor out the current `same-txid-diff-wtxid` transaction in `mempool_accept_wtxid.py` so that it can be reused.

laanwj commented on test/functional/test_framework/key.py:
```diff
@@ -181,6 +181,8 @@ def sign_ecdsa(self, msg, low_s=True, rfc6979=False):
         s = (pow(k, -1, ORDER) * (z + self.secret * r)) % ORDER
         if low_s and s > secp256k1.GE.ORDER_HALF:
```
Could use `elif`, but as the effect is the same, better to roll this into one `if`, i think:
```python
if (low_s and s > secp256k1.GE.ORDER_HALF) or (not low_s and s < secp256k1.GE.ORDER_HALF):
    s = ORDER - s
```

laanwj replied:
Right it's fine, it's documented well enough now, if anyone needs that functionality in the future they can add it then.

vasild commented on test/functional/p2p_segwit.py:
```diff
@@ -2066,6 +2069,89 @@ def received_wtxidrelay():
+        spending_tx = create_tx(funding_p2wpkh_tx.sha256, 0, amount - 1000, script_pubkey)
+        pubkey_hash = hash160(pubkey)
+        p2pkh_script = keyhash_to_p2pkh_script(pubkey_hash)
+        sign_witness_input(spending_tx, 0, p2pkh_script, amount, low_s=False)
```
I can imagine that malleating a transaction does not require the private key and a re-sign.

Currently there is this function:

```py
def create_malleated_version(self, tx)
```

which fills garbage. Could it be extended to take a boolean argument, indicating whether to fill with garbage like now, or flip `s` to `ORDER - s`? For this it would have to extract `s` like done on line 97 here:

https://github.com/bitcoin/bitcoin/blob/baa848b8d38187ce6b24a57cfadf1ea180209711/test/functional/test_framework/key.py#L59-L97

flip it to `ORDER - s` and re-do the DER encoding like:

https://github.com/bitcoin/bitcoin/blob/baa848b8d38187ce6b24a57cfadf1ea180209711/test/functional/test_framework/key.py#L184-L189

For the caller it would be useful to know if the new flipped `s` is low or high. I guess this can be indicated somehow by `create_malleated_version()` or there could be a separate function which takes a signed transaction and returns whether `s` is low or high.

vasild commented on test/functional/p2p_segwit.py:
```diff
@@ -2066,6 +2069,89 @@ def received_wtxidrelay():
+        test_transaction_acceptance(self.nodes[1], self.test_node, spending_tx, with_witness=True, accepted=True)
+        self.generate(self.nodes[0], 1)
+
+        # Example 2: P2WSH 1-of-2 multisig transaction where each signature applied to tx results in different wtxid
```
Isn't multisig kind of out-of-scope for creating a valid malleated transaction? I don't mind some extra tests, but those might make this PR more difficult to review. Maybe omit those if there is no review interest for some time.

vasild replied:
Now I see! Multisig is another way to create "same_txid, different_wtxid" even though it does not fiddle with the `s` value. And so, both transactions are relayable. Cool. Best would be to have a function that produces those two transactions. That would need the private keys. I guess better be separate from `create_malleated_version()`.

vasild reviewed (commented):
Concept ACK, thank you for looking into this!

theStack commented:
Concept ACK

Seems useful to create a pair of valid "same-txid-different-wtxid" transactions for testing purposes.

Didn't try that yet, but I think another possibility to create these without having to involve signatures at all would be a "trivial math puzzle"-like locking script of e.g. "OP_ADD OP_EQUAL 5" (in a p2wsh or p2tr script-path) and swapped witness data elements ([1,4] and [4,1]) in the spending txs.

stratospher committed:
test: refactor out same-txid-diff-wtxid tx to reuse in other tests

useful to easily create transactions with same txid, different
wtxid and valid witness for testing scenarios in other places
(ex: private broadcast connections)

theStack reviewed (approved):
ACK afaaba69eddd50bc22b94ca7c0f9195773aaa111

vasild reviewed (approved):
ACK afaaba69eddd50bc22b94ca7c0f9195773aaa111

The code seems to do what is intended and I confirm that it works. I guess it is useful as a generic tool to create a new transaction and two new children of it that have the same txid and different wtxid.

That is useful for the tests of https://github.com/bitcoin/bitcoin/pull/29415 too. However those tests already have a transaction and it would be easier for them and less code, if `malleate_tx()` could create a malleated/changed version of an existent transaction with a valid witness. See the [comment below](https://github.com/bitcoin/bitcoin/pull/32385#discussion_r2163643282).

Thanks!

vasild commented:
This is [used](https://github.com/bitcoin/bitcoin/pull/29415#issuecomment-3248422237) now in https://github.com/bitcoin/bitcoin/pull/29415, thank you!
